### ISSUE TITLE
LC0045 - Exceptions to the rule?

### ISSUE THREAD

**Issue by rvanbekkum** [StefanMaron/BusinessCentral.LinterCop#486]
We already have some enum objects where ID 0 is used for values with the name `Undefined`, `None`, and/or with `Caption = '', Locked = true;` (with the latter resulting in the enum value not showing as a selectable option in the BC client).

Could it be considered that a LC0045 diagnostic is not raised in this case as the developer deliberately thought about the value with ID 0?

**Comment by Arthurvdv** [comment]
The examples you're referring to are indeed valid to assign to the zero ID. I could use your help on how we should change the logic of this rule to support these scenario's.

At the moment the rule will verify that the Name (and Caption) need to be Empty. Applies a `.Trim()` to allow a white spaces.

* `value(0; "") { }`
* `value(0; "") { Caption = ''; }`

I'm unsure on howto allow cases like these below, without breaking the consistency of the rule.

* `value(0; "Undefined") { }`
* `value(0; "None") { }`
* `value(0; "") { Caption = 'Undefined'; }`

**Comment by rvanbekkum** [comment]
I think something like

```al
value(0; Undefined)
{
    Caption = '', Locked = true;
}
```

should be allowed.

I think for some of the checks you could reuse things that were done in https://github.com/StefanMaron/BusinessCentral.LinterCop/blob/master/Design/Rule0014PermissionSetCaptionLength.cs, like checking for `Locked = true`.

And/or for retrieving the `Name` of an enum value you could use `IEnumValueSymbol.Name`.

But maybe I am misunderstanding what you are asking? 😅

**Comment by Arthurvdv** [comment]
That was not what I meant 😅

I'm looking for what defines a Enum that is Undefined/None intent? At the moment the rule expects Name (and Caption) need to be Empty. That makes the rule easy to understand.

Do we hardcode that next to Empty also the values Undefined and None are acceptable?
Not sure how long this list of "Allowed Names" is going to be and where we draw the line of acceptable values. 🤔

For example something like this;

```AL
value(0; Empty)
{
    Caption = 'Empty';
}

**Comment by rvanbekkum** [comment]
I think there could be 2 things that we could address:

1. If the enum value has exactly `Caption = '', Locked = true` then the diagnostic should not be raised (because the value won't show up as a selectable option in the client)
2. A list of Names that should be allowed `None`, `Undefined` and maybe others that could be configured in LinterCop.json.

**Comment by Arthurvdv** [comment]
Hi, over the past months I’ve been putting a lot of time and care into the successor of BusinessCentral.LinterCop: ALCops (https://alcops.dev/). Because of that, this repository is no longer actively maintained.

I’m closing this issue, as it won’t be handled here anymore. If this is still relevant for you, please have a look at ALCops/Analyzers (https://github.com/ALCops/Analyzers).

To make the transition easier, you can also follow the LinterCop Migration Guide: https://alcops.dev/docs/lintercop-migration/

If you have any questions or run into issues during the transition, feel free to reach out through ALCops. Happy to help 🙂
